// trigger.dev/dist/esm/entryPoints/managed/logger.js
import { flattenAttributes } from "@trigger.dev/core/v3";
import { SimpleStructuredLogger } from "@trigger.dev/core/v3/utils/structuredLogger";
export class ManagedRunLogger {
    httpClient;
    env;
    logger;
    constructor(opts) {
        this.httpClient = opts.httpClient;
        this.env = opts.env;
        this.logger = new SimpleStructuredLogger("managed-run-logger");
    }
    sendDebugLog({ runId, message, date, properties, print = true }) {
        if (!runId) {
            runId = this.env.TRIGGER_RUN_ID;
        }
        if (!runId) {
            return;
        }
        const mergedProperties = {
            ...properties,
            runId,
            runnerId: this.env.TRIGGER_RUNNER_ID,
            workerName: this.env.TRIGGER_WORKER_INSTANCE_NAME,
        };
        if (print) {
            this.logger.log(message, mergedProperties);
        }
        const flattenedProperties = flattenAttributes(mergedProperties);
        this.httpClient.sendDebugLog(runId, {
            message,
            time: date ?? new Date(),
            properties: flattenedProperties,
        });
    }
}
export class ConsoleRunLogger {
    print;
    constructor(opts = {}) {
        this.print = opts.print ?? true;
    }
    sendDebugLog({ message, properties }) {
        if (this.print) {
            console.log("[ConsoleLogger]", message, properties);
        }
    }
}
