// @trigger.dev/core/dist/esm/v3/schemas/query.js
import { z } from "zod";
/**
 * Request body schema for executing a query
 */
export const QueryExecuteRequestBody = z.object({
    query: z.string(),
    scope: z.enum(["organization", "project", "environment"]).default("environment"),
    period: z.string().nullish(),
    from: z.string().nullish(),
    to: z.string().nullish(),
    format: z.enum(["json", "csv"]).default("json"),
});
/**
 * Response body schema for JSON format queries
 */
export const QueryExecuteJSONResponseBody = z.object({
    format: z.literal("json"),
    results: z.array(z.record(z.any())),
});
/**
 * Response body type for CSV format queries
 */
export const QueryExecuteCSVResponseBody = z.object({
    format: z.literal("csv"),
    results: z.string(),
});
export const QueryExecuteResponseBody = z.discriminatedUnion("format", [
    QueryExecuteJSONResponseBody,
    QueryExecuteCSVResponseBody,
]);
// ---------------------------------------------------------------------------
// Query table row types
// ---------------------------------------------------------------------------
/**
 * User-facing friendly run status values returned by the query system.
 */
export const runFriendlyStatus = [
    "Delayed",
    "Queued",
    "Pending version",
    "Dequeued",
    "Executing",
    "Waiting",
    "Reattempting",
    "Paused",
    "Canceled",
    "Interrupted",
    "Completed",
    "Failed",
    "System failure",
    "Crashed",
    "Expired",
    "Timed out",
];
