// @trigger.dev/schema-to-json/dist/commonjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.schemaToJsonSchema = schemaToJsonSchema;
exports.canConvertSchema = canConvertSchema;
exports.isZodSchema = isZodSchema;
const zod_to_json_schema_1 = require("zod-to-json-schema");
const z4 = __importStar(require("zod/v4"));
const yup_to_json_schema_1 = require("@sodaru/yup-to-json-schema");
const effect_1 = require("effect");
/**
 * Convert a schema from various validation libraries to JSON Schema
 *
 * This function attempts to convert schemas without requiring external dependencies to be bundled.
 * It will only succeed if:
 * 1. The schema has built-in JSON Schema conversion (ArkType, Zod 4, TypeBox)
 * 2. The required conversion library is available at runtime (zod-to-json-schema, @sodaru/yup-to-json-schema, etc.)
 *
 * @param schema The schema to convert
 * @param options Optional conversion options
 * @returns The conversion result or undefined if conversion is not possible
 */
function schemaToJsonSchema(schema, options) {
    const parser = schema;
    // Check if schema has a built-in toJsonSchema method (e.g., ArkType, Zod 4)
    if (typeof parser.toJsonSchema === "function") {
        try {
            const jsonSchema = parser.toJsonSchema();
            return {
                jsonSchema,
            };
        }
        catch (error) {
            // If toJsonSchema fails, continue to other checks
        }
    }
    if (isZodSchema(parser)) {
        const jsonSchema = convertZodSchema(parser, options);
        if (jsonSchema) {
            return {
                jsonSchema: jsonSchema,
            };
        }
    }
    // Check if it's a TypeBox schema (has Static and Kind symbols)
    if (parser[Symbol.for("TypeBox.Kind")] !== undefined) {
        // TypeBox schemas are already JSON Schema compliant
        return {
            jsonSchema: parser,
        };
    }
    if (isYupSchema(parser)) {
        const jsonSchema = convertYupSchema(parser);
        if (jsonSchema) {
            return {
                jsonSchema: jsonSchema,
            };
        }
    }
    if (isEffectSchema(parser)) {
        const jsonSchema = convertEffectSchema(parser);
        if (jsonSchema) {
            return {
                jsonSchema: jsonSchema,
            };
        }
    }
    // Future schema types can be added here...
    // Unknown schema type
    return undefined;
}
/**
 * Check if a schema can be converted to JSON Schema
 */
function canConvertSchema(schema) {
    const result = schemaToJsonSchema(schema);
    return result !== undefined;
}
function isZodSchema(schema) {
    return isZod3Schema(schema) || isZod4Schema(schema);
}
function isZod3Schema(schema) {
    return "_def" in schema && "parse" in schema && "parseAsync" in schema && "safeParse" in schema;
}
function isZod4Schema(schema) {
    return "_zod" in schema;
}
function convertZodSchema(schema, options) {
    if (isZod4Schema(schema)) {
        return convertZod4Schema(schema, options);
    }
    if (isZod3Schema(schema)) {
        return convertZod3Schema(schema, options);
    }
    return undefined;
}
function convertZod3Schema(schema, options) {
    const useReferences = options?.useReferences ?? false;
    return (0, zod_to_json_schema_1.zodToJsonSchema)(schema, {
        $refStrategy: useReferences ? "root" : "none",
    });
}
function convertZod4Schema(schema, options) {
    const useReferences = options?.useReferences ?? false;
    return z4.toJSONSchema(schema, {
        target: "draft-7",
        io: "output",
        reused: useReferences ? "ref" : "inline",
    });
}
function isYupSchema(schema) {
    return "spec" in schema && "_typeCheck" in schema;
}
function convertYupSchema(schema) {
    try {
        return (0, yup_to_json_schema_1.convertSchema)(schema);
    }
    catch {
        return undefined;
    }
}
function isEffectSchema(schema) {
    return "ast" in schema && typeof schema.ast === "object" && typeof schema.ast._tag === "string";
}
function convertEffectSchema(schema) {
    try {
        return effect_1.JSONSchema.make(schema);
    }
    catch {
        return undefined;
    }
}
